pkg/sql/sem/tree/type_name.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tree
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/errors"
+)
+
+// TypeReferenceResolver is the interface that will provide the ability
+// to actually lookup type metadata and transform references into
+// *types.T's. In practice, this will probably be implemented by
+// the planner, but for now it is a dummy interface.
+type TypeReferenceResolver interface {
+	ResolveType() (*types.T, error)
```

Comment:
I don't get this. The entire catalog is a Resolver, so what does `ResolveType()` return? The type of what? I would expect it to take some argument..

Reply:
Sorry, this is still unfinished. I embedded this resolver into different structs to iterate quicker. The end goal is for the `planner` to implement this interface, and pass itself down to the catalog, semaCtx and evalCtx where needed.

I don't know yet what exactly `ResolveType` needs to take in -- once more progress is made on how we are going to perform name resolution of types this will be more clear. As of this PR we don't have any unresolved type names created anywhere yet, so this unused. I want this to look something the corresponding functions on the `tableName` struct that perform name resolution and return descriptors.